libtransmission/blocklist.cc

```diff
@@ -492,17 +495,18 @@ std::optional<Blocklist> Blocklist::saveNew(std::string_view external_file, std:
     auto const src_file = std::string{ std::data(bin_file), std::size(bin_file) - std::size(BinFileSuffix) };
     tr_sys_path_remove(src_file.c_str());
     tr_error* error = nullptr;
-    if (!tr_sys_path_copy(tr_pathbuf{ external_file }, src_file.c_str(), &error))
+    auto const copied = tr_sys_path_copy(tr_pathbuf{ external_file }, src_file.c_str(), &error);
+    if (error != nullptr)
+    {
+        tr_logAddWarn(fmt::format(
+            _("Couldn't save '{path}': {error} ({error_code})"),
+            fmt::arg("path", src_file),
+            fmt::arg("error", error->message),
+            fmt::arg("error_code", error->code)));
+        tr_error_clear(&error);
+    }
+    if (!copied)
```

Comment:
These 2 conditions check the same thing: if return value is `false`, `error` must be non-null (if passed in); if return value is `true`, `error` must be null. If anything, I think the `error != nullptr` check should've been removed and return value check should've stayed, but I don't know what the reported issue was about. I'm saying return value check is the right thing to do because that's the general approach due to error out parameter being optional (so functions always use other means to indicate error if not passed); if we make it required (e.g. how some Boost functions provide two overloads, throwing one and one accepding and out error code), then returning `bool` won't even make sense.

Reply:
> If anything, I think the error != nullptr check should've been removed and return value check should've stayed

That would just trigger the inverse warning, that we were using `error` without checking its value first.

You're 100% right, the root problem is that we have two ways of checking for error in these functions that take a `tr_error*` -- the return value, and the error structure. What I did in this PR was to split the diference byusing `error != nullptr` just for error reporting, and `copied` for control flow.

IMO the right answer would be to modify the functions to reduce the "two ways to check for success" down to "one way to check for success" but that refactor is too big for a cleanup PR.